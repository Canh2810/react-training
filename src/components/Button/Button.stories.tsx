import { StoryObj, Meta } from '@storybook/react'

import Button from './'
import { ButtonVariants } from '@types'
import { Search } from '@assets/icons'

export default {
  title: 'Components/Button',
  component: Button,
} as Meta

type Story = StoryObj<typeof Button>

export const Default: Story = {
  args: {
    variant: ButtonVariants.Outlined,
    children: 'Click me',
    onClick: () => {
      alert('Button component')
    },
  },
}

export const Outlined: Story = {
  args: {
    variant: ButtonVariants.Container,
    children: 'Click me',
    onClick: () => {
      alert('Button component')
    },
  },
}

export const IconButton: Story = {
  args: {
    children: <Search />,
    onClick: () => {
      alert('Button component')
    },
  },
}
