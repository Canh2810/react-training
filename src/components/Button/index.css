.btn {
  padding: 12px 20px;
  border-radius: 6px;
  border: none;
  background-color: unset;
  cursor: pointer;
}

.btn:hover {
  opacity: 0.7;
}
