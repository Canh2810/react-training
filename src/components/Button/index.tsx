// Libs
import { memo, ButtonHTMLAttributes, ReactNode } from 'react'

// Types
import { ButtonVariants } from '@types'

// Themes
import { variants } from '@themes'

// Stores
import { useThemeStore } from '@stores'

// Styles
import './index.css'

export interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  variant?: ButtonVariants
  children?: ReactNode
  isLoading?: boolean
  className?: string
  onClick?: () => void
}

const Button = ({
  variant = ButtonVariants.Outlined,
  children,
  isLoading,
  className = '',
  onClick,
  ...props
}: ButtonProps) => {
  const [theme] = useThemeStore((state) => [state.theme])

  const buttonStyle =
    variant === ButtonVariants.Container
      ? variants.container
      : variants.outlined(theme)

  return (
    <button
      className={`btn ${className}`}
      disabled={isLoading}
      onClick={onClick}
      style={{ ...buttonStyle }}
      {...props}
    >
      {children}
    </button>
  )
}

export default memo(Button)
