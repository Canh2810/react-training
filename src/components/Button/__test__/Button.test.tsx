import { fireEvent, render } from '@testing-library/react'
import Button from '..'
import { ButtonVariants } from '@types'

describe('Button component', () => {
  it('should render with outlined variant', () => {
    const { container } = render(<Button>Button</Button>)
    expect(container).toMatchSnapshot()
  })

  it('should render with container variant', () => {
    const { container } = render(
      <Button variant={ButtonVariants.Container}>Button</Button>,
    )
    expect(container).toMatchSnapshot()
  })

  it('handles onClick event', () => {
    const onClickMock = jest.fn()
    const { getByText } = render(
      <Button onClick={onClickMock}>Click me</Button>,
    )
    const button = getByText('Click me')
    fireEvent.click(button)
    expect(onClickMock).toHaveBeenCalledTimes(1)
  })

  it('disables the button when isLoading is true', () => {
    const { getByRole } = render(<Button isLoading={true}>Loading</Button>)
    const button = getByRole('button')
    expect(button).toBeDisabled()
  })
})
