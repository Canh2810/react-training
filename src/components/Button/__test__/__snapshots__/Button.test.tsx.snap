// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component should render with container variant 1`] = `
<div>
  <button
    class="btn "
    style="background-color: rgb(75, 107, 251); color: rgb(255, 255, 255);"
  >
    Button
  </button>
</div>
`;

exports[`Button component should render with outlined variant 1`] = `
<div>
  <button
    class="btn "
    style="border: 1px solid; color: rgb(105, 106, 117); border-color: rgba(105, 106, 117, 0.30); background-color: rgb(255, 255, 255);"
  >
    Button
  </button>
</div>
`;
